### ISSUE TITLE
[RFC] `out` params: allows to call in-place `proc` with rvalues

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8808]
## motivating example
The following pattern is typical:
```nim
proc normalizePath*(path: var string) = ...
proc normalizedPath*(path: string): string =
  ## Returns a normalized path for the current OS. See `<#normalizePath>`_
  result = path
  normalizePath(result)
```
the 1st `proc` allows efficiency by working inplace
the 2nd `proc` is more convenient in lots of cases, eg when working with RVALUES or in UFCS chains, eg:

```nim
let path = getConfig().normalizePath
# instead of:
var s=getConfig() # awkward: now `s` is `var` even though we'd maybe want it `let`
normalizePath(s) 
```

## proposal for this concrete example 
With this proposal, we'd simply write the 1st version, using `out` instead of `var`:
```nim
proc normalizePath*(path: out string) = ...
# this implicitly generates both version1 and version2 above, giving use the best of both worlds without all the boilerplate.
```

## proposal details: general case
* a proc `foo` returning `void` is allowed to have a single `out` param:
```nim
proc foo*(argsBefore..., ret: out T, argsAfter...)=...
```
when there's such an `out` param, it generates an implicit proc:
```nim
proc foo*(argsBefore..., ret: var T, argsAfter...): T =
  result = ret
  foo(argsBefore..., result, argsAfter...)
```

* the pragmas and visibility (eg `*`) are same as original `foo`.
* can also be extended to work with `template` instead of `proc`
* there can be any number (including 0) of arguments before and after the out param
* the `out` param is allowed to have a default value; nothing special in this case: the implicit proc will also have that default value for `ret`, eg:
```nim
proc foo(n: int, s: out = "") = ...
# generates:
proc foo(n: int, s = "") = ...
# these work, nothing special:
var s = "foo"
foo(10, s)
let s2 = foo(10, "foo")
let s2 = foo(10)
```

* special case of `openArray`
for `openArray[T]`, `result = ret` is not legal, but simply make return type a seq[T]:
```nim
proc foo*(argsBefore..., ret: var T, argsAfter...): seq[T] =
  result = ret.toSeq # easy to implement; or see https://github.com/nim-lang/Nim/pull/8711
  foo(argsBefore..., result, argsAfter...)
```
Note: this is exactly what's being done (manually...) in many places, eg: `sort` vs `sorted` boil down to exactly that.

With this proposal, we'd simply write:
`proc sort*[T](a: out openArray[T], cmp: proc (x, y: T): int {.closure.}, order = SortOrder.Ascending)`
and we don't need the `sorted` variant, which can be deprecated.

## benefits
* leads to simpler and more uniform API's (no need for boilerplate to make a proc work both inplace and in UFCS)
* avoids ad-hoc naming conventions, eg: `reverse` vs `reversed`, `sort` vs `sorted` etc
* allows a `out` variable to be `let`, which is nice:
```nim
proc foo(s: var string, n: 10)
# this is `var` because `foo` takes `s` by var
var s = "hello"
foo(s, 10)

proc foo2(s: out string, n: 10)
let s = foo2("hello", 10) # `s` can be declared as `let`!
```

* no need to choose whether we should allow UFCS or in-place processing for a proc: just use `out` where applicable

## Note
* this feature exploits the fact that Nim doesn't have implicit `discard`, which removes all ambiguity
* `out` is just an idea; feel free to suggest a different name

## how often is that useful?
very often: pretty much every time you have a `proc` that returns `void` and has 1 of its parameters marked as `var`.

cases where we (currently) only have the inplace version, eg:
```nim
proc removeSuffix*(s: var string, suffix: string) # => could be more convenient with an `out` param
```

cases where we have both version1 and version2, eg:
```nim
proc escapeJsonUnquoted*(s: string; result: var string)
proc escapeJsonUnquoted*(s: string): string
```
there are a bunch of these.
In particular when the `var` param is an openArray, it'll simplify code a lot.

**Comment by timotheecour** [comment]
right, I thought of the alternative approach after posting the rest; if it can be done (looks like it may), it may be the simplest. `outparam` would be a worthy macro.

**Comment by hlaaftana** [comment]
Alternative: always define the version with the var string, and add a new template called tapVar or whatever name that you can use at the callsite.

```nim
template tapVar(val, body: untyped): untyped =
  var it {.inject.} = val
  body
  it

echo tapVar("abcd", removeSuffix(it, "ab"))
```

If it's too verbose to use it on every call, define overloads with it.

```
template escapeJsonUnquoted(s: string): string =
  tapVar(newStringOfCap(s.len + s.len shr 3), escapeJsonUnquoted(it, s))
```

Yes I know the name is horrible, so maybe a macro for defining overloads is a better solution

**Comment by Araq** [comment]
My own proposal for that is like this: There is a yet-another rewrite rule in overloading resolution: 

A proc `p` that is of a signature like

```nim

proc p(args...; result: var T) # note: parameter name must be the last and be named ``result``

```

can be called as ``p(args)`` and this call is rewritten into ``(var tmp: T; p(args, tmp); tmp)``. I deprecated the "self/this" rewrite rule in order to allow for this rule instead which I think is much more useful and inline with Nim's goal that efficient code should be easy to write.

**Comment by Araq** [comment]
> No keyword introduced, but might be confusing to newcomers. Maybe writing it like this will make it clearer:

I think you misunderstand my proposal. ``p``'s declaration looks like before and even in today's Nim ``result`` has some special behaviour.

> result is already an invalid proc parameter name and so its type is always guaranteed to be var T in a proc body where the proc returns T. 

No, in ``proc p(): T`` its return type is ``T``, not ``var T``.

> I don't see why it should be the last parameter though. It could act like any comma before or after it is a semicolon so it doesn't break using.

Because it's easiest to implement this way.

**Comment by hlaaftana** [comment]
> I think you misunderstand my proposal. p's declaration looks like before and even in today's Nim result has some special behaviour.

Yes, result having special behaviour already means you shouldn't have to type out `result: var T`. The difference with the new declaration is a newcomer can always understand what the return type is without having to know that it's in the proc parameters. Also you can choose whether `p` returns `T` and `var T` but I can't think of a situation where that'd work or have merit.

> No, in proc p(): T its return type is T, not var T.

I meant the type of the name `result`.

**Comment by Araq** [comment]
> Isn't that possible without any special compiler support? If so, it's probably the prefererad solution. IMO the benefit of this feature (avoiding writing a three line wrapper proc) is to small to motivate a compiler feature.

But the macro solution would produce 2 procs out of one, these procs end up in the generated docs and in the error messages. Might still be good enough though.

**Comment by Araq** [comment]
> I meant the type of the name result.

You lost me. Do you only dislike the fact that this parameter needs to be named ``result`` in my proposal? Or is it something else.

**Comment by hlaaftana** [comment]
My point was that the type annotation in `result: var T` was redundant, I could've expressed it better.

**Comment by Araq** [comment]
How is it redundant? The return type is ``void``.

**Comment by timotheecour** [comment]
@araq 
your proposal [here](https://github.com/nim-lang/Nim/issues/8808#issuecomment-417289979) 
(of allowing to call `proc p(args...; result: var T)` with `p(args)`) is exactly what I had suggested in D [here](https://digitalmars.com/d/archives/digitalmars/D/Ouput-UFCS_proposal_void_fun_A_a_ref_B_b_B_fun_A_a_188566.html) a while ago; in Nim I realized we can do something much more useful (while still allowing this as a special case) thanks to Nim's feature of no `implicit discard`, see below:

with your proposal, you're not addressing the `motivating example` in original post:
```nim
proc normalizePath*(result: var string) = ...
# you can only do the following, which is not very useful
let foo = normalizePath()

# or the usual way:
var s = someExpr()
normalizePath(s)
```

with my proposal (regardless of whether we use `out` or `{. outparam .}`), we have:
```
proc normalizePath*(path: var string) = {. outparam .} ...
let foo = someExpr().normalizePath
# or the usual way [...]
```

Furthermore, **my proposal allows everything allowed by your proposal** (as I had mentioned in original post, see example under `these work, nothing special:`):
```
proc p*(foo: var string = "") = {. outparam .} ...
let foo = p()

# more generally:
proc p*[T](args1..., foo: var T = default(T), args2....) = {. outparam .} ...
let foo = p(args1..., args2....)
```
# note: for `default(T)`, see my PR https://github.com/nim-lang/Nim/pull/8490 (marked as postponed, but basically it works)

**Comment by Araq** [comment]
1. Nothing in my proposal truely requires that the parameter's name shall be ``result``, we can remove this constraint (though I think it makes lots of sense).
2. Your proposal of either allowing the first parameter as the output parameter or an arbitrary one seems vastly harder to implement. 
3. It's true that my proposal doesn't work that well with dot-chaining, yours however doesn't give us **clean** "append semantics".
4. Your transformation with ``result = ret`` introduces a copy and seems to create the "faster" version by calling the slower variant!
